pkg/didcomm/protocol/route/service.go

```diff
@@ -133,18 +146,98 @@ func New(prov provider) (*Service, error) {
 		connectionLookup:     connectionLookup,
 		routeRegistrationMap: make(map[string]chan Grant),
 		keylistUpdateMap:     make(map[string]chan *KeylistUpdateResponse),
-	}, nil
+		callbacks:            make(chan *callback),
+	}
+
+	go s.listenForCallbacks()
+
+	return s, nil
+}
+
+func (s *Service) listenForCallbacks() {
+	for c := range s.callbacks {
+		logger.Debugf("handling user callback for msgID=%s err=%s", c.msg.ID(), c.err)
+
+		if c.err != nil {
+			go s.handleUserRejection(c)
+			continue
+		}
+
+		switch c.msg.Type() {
+		case RequestMsgType:
+			err := s.handleInboundRequest(c)
+			if err != nil {
+				logger.Errorf("failed to handle inbound request: %+v : %w", c.msg, err)
+			}
+		default:
+			logger.Warnf("ignoring unsupported message type %s", c.msg.Type())
+		}
+	}
+}
+
+func (s *Service) handleUserRejection(c *callback) {
+	logger.Infof("user aborted response action for msgID=%s", c.msg.ID())
+}
+
+func triggersActionEvent(msgType string) bool {
+	return msgType == RequestMsgType
+}
+
+func (s *Service) sendActionEvent(msg service.DIDCommMsg, myDID, theirDID string) error {
+	events := s.ActionEvent()
+	if events == nil {
+		return fmt.Errorf("no clients registered to handle action events for %s protocol", Coordination)
+	}
+
+	logger.Debugf("dispatching action event for msg=%+v myDID=%s theirDID=%s", msg, myDID, theirDID)
+
+	go func() {
+		c := &callback{
+			msg:      msg,
+			myDID:    myDID,
+			theirDID: theirDID,
+		}
+
+		events <- service.DIDCommAction{
+			ProtocolName: Coordination,
+			Message:      msg,
+			Continue: func(args interface{}) {
+				opts, ok := args.(Options)
+				if !ok {
+					opts = Options{}
+				}
+
+				c.options = &opts
+
+				s.callbacks <- c
+			},
+			Stop: func(err error) {
+				c.err = err
+
+				s.callbacks <- c
+			},
+		}
+	}()
+
+	return nil
 }
 
 // HandleInbound handles inbound route coordination messages.
 func (s *Service) HandleInbound(msg service.DIDCommMsg, myDID, theirDID string) (string, error) { // nolint gocyclo (5 switch cases)
+	logger.Debugf("input: msg=%+v myDID=%s theirDID=%s", msg, myDID, theirDID)
+
+	if triggersActionEvent(msg.Type()) {
+		err := s.sendActionEvent(msg, myDID, theirDID)
```

Comment:
`return s.sendActionEvent(msg, myDID, theirDID)` or wrap the error

Reply:
done